Simplified/AxisService/NYPLAxisXML.swift

```diff
@@ -0,0 +1,125 @@
+//
+//  AxisXML.swift
+//  Simplified
+//
+//  Created by Raman Singh on 2021-04-06.
+//  Copyright © 2021 NYPL Labs. All rights reserved.
+//
+
+import Foundation
+
+#if AXIS
+
+/// A wrapper class which aids debugging and allows recursive search with a given key in an NYPLXML object.
+struct NYPLAxisXML: CustomStringConvertible {
+  let children: [NYPLAxisXML]
+  let attributes: [AnyHashable: Any]
+  let name: String
+  let nameSpaceURI: String
+  let qualifiedName: String
+  let value: String
+  
+  // MARK: - Static Constants
+  static private let summary = "AXIS: Failed to generate AXISXML from data"
+  static let NYPLXMLGenerationFailure = "Failed generating NYPL XML from data"
+  
+  init(xml: NYPLXML) {
+    self.children = xml.children.compactMap({ $0 as? NYPLXML }).map { NYPLAxisXML(xml: $0) }
+    self.attributes = xml.attributes
+    self.name = xml.name
+    self.nameSpaceURI = xml.namespaceURI
+    self.qualifiedName = xml.qualifiedName
+    self.value = xml.value
+  }
+  
+  /// Failable initializer for AXISXML. Returns nil if data is nil or NYPLXML initialization fails. Logs erron on
+  /// failure
+  init?(data: Data?) {
+    guard let data = data else {
+      NYPLErrorLogger.logError(
+        withCode: .axisDRMFulfillmentFail,
+        summary: NYPLAxisXML.summary,
+        metadata: [NYPLAxisService.reason: NYPLAxisService.nilDataFromFileURLFailure])
+      
+      return nil
+    }
+    
+    guard let xml = NYPLXML(data: data) else {
+        NYPLErrorLogger.logError(
+          withCode: .axisDRMFulfillmentFail,
+          summary: NYPLAxisXML.summary,
+          metadata: [NYPLAxisService.reason: NYPLAxisXML.NYPLXMLGenerationFailure])
+        
+        return nil
+    }
+    
+    self.init(xml: xml)
+  }
+  
+  /// For debugging purposes
+  var description: String {
+    var result = "children: \(children.map { $0.name })"
+    result = result + "\nattributes: \(attributes)"
+    result = result + "\nname: \(name)"
+    result = result + "\nnameSpaceURI: \(nameSpaceURI)"
+    result = result + "\nqualifiedName: \(qualifiedName)"
+    result = result + "\nvalue: \(value)"
+    return result
+  }
+  
+  /// Finds and returns all values within the xml and its children matching the given key.
+  ///
+  /// Note: We're doing something akin to a `depth first pre-order search` of a general tree here.
+  /// The time complexity is `O(n)` where n is the number of nodes (children and their children and so forth).
+  /// The space complexity will be `O(|n|)` where n is the average size of each node.
```

Comment:
shouldn't `n` refer to the number of nodes? usually that's the case.... otherwise this would be `n x m` where `m` is the avg size of node, but that's constant.